torch/jit/_script.py

```diff
@@ -1079,5 +1080,100 @@ def _unwrap_optional(x):
+        _jit_internal.is_tuple(target_type)
+    elif target_type is Dict or target_type is dict:
+        _jit_internal.is_dict(target_type)
+    elif target_type is None or target_type is Optional:
+        _jit_internal.is_optional(target_type)
+
+# supports List/Dict/Tuple and Optional types
+# TODO support future
+def generics_checker(obj, target_type):
+    origin_type = get_origin(target_type)
+    check_args_exist(target_type)
+    if origin_type is list or origin_type is List:
+        if not isinstance(obj, list):
+            return False
+        arg_type = get_args(target_type)[0]
+        arg_origin = get_origin(arg_type)
+        for el in obj:
+            # check if nested generics, ex: List[List[str]]
+            if arg_origin:  # processes nested generics, ex: List[List[str]]
+                if not generics_checker(el, arg_type):
+                    return False
+            elif not isinstance(el, arg_type):
+                return False
+        return True
+    elif origin_type is Dict or origin_type is dict:
+        if not isinstance(obj, dict):
+            return False
+        key_type = get_args(target_type)[0]
+        val_type = get_args(target_type)[1]
+        for key, val in obj.items():
+            # check if keys are of right type
+            if not isinstance(key, key_type):
+                return False
+            val_origin = get_origin(val_type)
+            if val_origin:
+                if not generics_checker(val, val_type):
+                    return False
+            elif not isinstance(val, val_type):
+                return False
+        return True
+    elif origin_type is Tuple or origin_type is tuple:
+        if not isinstance(obj, tuple):
+            return False
+        arg_types = get_args(target_type)
+        if len(obj) != len(arg_types):
+            return False
+        for el, el_type in zip(obj, arg_types):
+            el_origin = get_origin(el_type)
+            if el_origin:
+                if not generics_checker(el, el_type):
+                    return False
+            elif not isinstance(el, el_type):
+                return False
+        return True
+    elif origin_type is Union:  # actually handles Optional Case
+        if obj is None:  # check before recursion because None is always fine
+            return True
+        optional_type = get_args(target_type)[0]
+        optional_origin = get_origin(optional_type)
+        if optional_origin:
+            return generics_checker(obj, optional_type)
+        elif isinstance(obj, optional_type):
+            return True
+    return False
+
+
+def _isinstance(obj, target_type) -> bool:
+    origin_type = get_origin(target_type)    
+    if origin_type:
+        return generics_checker(obj, target_type)
+
+    # handle python 3.6 quality of returning None for origin of 
+    # containers without contained type (vs returning container type)
+    if version_info[1] == 6:
```

Comment:
```suggestion
    check_args_exist(target_type) 
```
Here I could just call `check_args_exist(target_type)` no matter what the version number and the code behaves the same